Sources/System/Sockets/SocketAddress.swift

```diff
@@ -21,35 +24,81 @@ public struct SocketAddress {
   }
 
   public init(_ buffer: UnsafeRawBufferPointer) {
-    precondition(buffer.count >= MemoryLayout<CInterop.SockAddr>.size)
-    if buffer.count <= MemoryLayout<_InlineStorage>.size {
+    self.init(unsafeUninitializedCapacity: buffer.count) { target in
+      target.baseAddress!.copyMemory(
+        from: buffer.baseAddress!,
+        byteCount: buffer.count)
+      return buffer.count
+    }
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketAddress {
+  /// Initialize an empty socket address with the specified minimum capacity.
+  /// The default capacity makes enough storage space to fit any IPv4/IPv6 address.
+  ///
+  /// Addresses with storage preallocated this way can be repeatedly passed to
+  /// `SocketDescriptor.receiveMessage`, eliminating the need for a potential
+  /// allocation each time it is called.
+  public init(minimumCapacity: Int = Self.defaultCapacity) {
+    self.init(unsafeUninitializedCapacity: minimumCapacity) { buffer in
+      system_memset(buffer, to: 0)
+      return 0
+    }
+  }
+
+  /// Reserve storage capacity such that `self` is able to store addresses
+  /// of at least `minimumCapacity` bytes without any additional allocation.
+  ///
+  /// Addresses with storage preallocated this way can be repeatedly passed to
+  /// `SocketDescriptor.receiveMessage`, eliminating the need for a potential
+  /// allocation each time it is called.
+  public mutating func reserveCapacity(_ minimumCapacity: Int) {
+    guard minimumCapacity > _capacity else { return }
+    let length = _length
+    var buffer = _RawBuffer(minimumCapacity: minimumCapacity)
+    buffer.withUnsafeMutableBytes { target in
+      self.withUnsafeBytes { source in
+        assert(source.count == length)
+        assert(target.count > source.count)
+        if source.count > 0 {
+          target.baseAddress!.copyMemory(
+            from: source.baseAddress!,
+            byteCount: source.count)
+        }
+      }
+    }
+    self._variant = .large(length: length, bytes: buffer)
+  }
+
+  public init(
```

Comment:
Add a doc comment, likely cribbing from the stdlib API. Or, should this be internal for now?